src/com/esotericsoftware/kryonet/rmi/ObjectSpace.java

```diff
@@ -404,11 +406,25 @@ public Object invoke (Object proxy, Method method, Object[] args) throws Excepti
 			boolean needsResponse = !udp && (transmitReturnValue || transmitExceptions || !nonBlocking);
 			byte responseID = 0;
 			if (needsResponse) {
-				synchronized (this) {
-					// Increment the response counter and put it into the low bits of the responseID.
-					responseID = nextResponseId++;
-					if (nextResponseId > responseIdMask) nextResponseId = 1;
-					pendingResponses[responseID] = true;
+				synchronized (invokeLocker) {
+					while (true) {
+						synchronized (this) {
+							if (totalPendingResponsesCnt < responseIdMask) {
+								// Find the first non-pending responseID.
+								do {
+									responseID = nextResponseId++;
+									if (nextResponseId > responseIdMask) nextResponseId = 1;
+								} while (pendingResponses[responseID]);
+								pendingResponses[responseID] = true;
+								totalPendingResponsesCnt++;
+								break;
+							}
+						}
+						//This sleep is under lock "synchronized(invokeLocker)" but not under lock "synchronized(this)"
+						//So all sequenced invocations will wait (and totalPendingResponsesCnt could not increased),
+						// but invocation result can be handled (and totalPendingResponsesCnt could decreased)
+						Thread.sleep(10);
```

Comment:
Library code cannot just sleep for an arbitrary amount of time! Thread.yeild would be better, but even that isn't very good.

Please explain what the ObjectSpace changes are about.